flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java

```diff
@@ -0,0 +1,354 @@
+				rmLeaderElectionService.isLeader(rmLeaderSessionId).get(timeout.toMilliseconds(), TimeUnit.MILLISECONDS);
+			}
+		}
+
+		/**
+		 * Start the resource manager and grant leadership to it.
+		 */
+		public void startResourceManager() throws Exception {
+			resourceManager.start();
+			rmServices.grantLeadership();
+		}
+
+		/**
+		 * Stop the Akka actor system.
+		 */
+		public void stopResourceManager() throws Exception {
+			rpcService.stopService();
+		}
+	}
+
+	static class TestContainer extends UtilsTest.TestingContainer {
+		Resource resource;
+		Priority priority;
+
+		TestContainer(String id, String host) {
+			super(id, host);
+		}
+
+		@Override
+		public Resource getResource() {
+			return resource;
+		}
+
+		@Override
+		public void setResource(Resource resource) {
+			this.resource = resource;
+		}
+
+		@Override
+		public Priority getPriority() {
+			return priority;
+		}
+
+		@Override
+		public void setPriority(Priority priority) {
+			this.priority = priority;
+		}
+	}
+
+	@Test
+	public void testStopWorker() throws Exception {
+		new Context() {{
+			startResourceManager();
+			// Request slot from SlotManager.
+			resourceManager.runInMainThread(() -> {
+				try {
+					rmServices.slotManager.registerSlotRequest(
+							new SlotRequest(new JobID(), new AllocationID(), resourceProfile1, taskHost));
+				} catch (SlotManagerException e) {
+					e.printStackTrace();
+				}
+			});
+
+			// Callback from YARN when container is allocated.
+			Container testingContainer = new TestContainer(task.toString(), taskHost);
+			testingContainer.setResource(Resource.newInstance(200, 1));
+			testingContainer.setPriority(Priority.UNDEFINED);
+			resourceManager.onContainersAllocated(ImmutableList.of(testingContainer));
+
+			// Remote task executor registers with YarnResourceManager.
+			TaskExecutorGateway mockTaskExecutorGateway = mock(TaskExecutorGateway.class);
+			rpcService.registerGateway(taskHost, mockTaskExecutorGateway);
+			resourceManager.registerTaskExecutor(taskHost,
+					new ResourceID(testingContainer.getId().toString()),
+					slotReport,
+					Time.seconds(10)).handleAsync(
+							(RegistrationResponse response, Throwable throwable) -> {
+								assertTrue(rmServices.slotManager.getNumberRegisteredSlots() == 1);
+								return null;
+							});
+
+			// Unregister all task executors and release all containers.
+			resourceManager.runInMainThread(() -> {
+				rmServices.slotManager.unregisterTaskManagersAndReleaseResources();
+			});
+			stopResourceManager();
+			verify(mockResourceManagerClient).addContainerRequest(any(AMRMClient.ContainerRequest.class));
+			verify(mockNMClient).startContainer(eq(testingContainer), any(ContainerLaunchContext.class));
+			verify(mockNMClient).stopContainer(any(ContainerId.class), any(NodeId.class));
+			verify(mockResourceManagerClient).releaseAssignedContainer(any(ContainerId.class));
+			assertTrue(rmServices.slotManager.getNumberRegisteredSlots() == 0);
```

Comment:
Maybe add a comment that it's now safe to access the `SlotManager` state since the `ResourceManager` has been stopped.